Added improvement in documentation for parallel streaming invocations limit (#37194)
dotnet/AspNetCore.Docs: aspnetcore/signalr/hubs.md

```

> [!NOTE]
> This feature makes use of <xref:Microsoft.Extensions.DependencyInjection.IServiceProviderIsService>, which is optionally implemented by DI implementations. If the app's DI container doesn't support this feature, injecting services into hub methods isn't supported.

> ## Limit per-connection streaming invocations

Reviewer: This new section would need to move to after the sub section _Keyed Services support in Dependancy Injection_.
"Keyed services support" is conceptually a subtopic of "Inject services into a hub"

So the organizaiton would like like this:

`## Inject services into a hub`
  (existing content)
`  ### Keyed services support in Dependency Injection`
    (existing content)

`## Limit per-connection streaming invocations`
  (new content)

`## Handle events for a connection`

`HubOptions.MaximumParallelInvocationsPerClient` controls non-streaming hub invocations only. It does not apply to streaming hub invocations. Streaming invocations are expected to be long-running and can run concurrently. To enforce a per-connection limit for streaming invocations, use a hub filter to track active stream-returning hub methods.

```csharp

> using System.Collections.Concurrent;

Reviewer: I think we also need:
`using System.Threading.Channels;`

public class StreamConcurrencyFilter : IHubFilter
{
    private readonly ConcurrentDictionary<string, int> _activeStreams = new();
    private readonly int _maxConcurrentStreams;

    public StreamConcurrencyFilter(int maxConcurrentStreams = 1)
    {
        _maxConcurrentStreams = maxConcurrentStreams;
    }

    public async ValueTask<object?> InvokeMethodAsync(
        HubInvocationContext invocationContext,
        Func<HubInvocationContext, ValueTask<object?>> next)
    {
        if (!IsStreamingInvocation(invocationContext.HubMethod.ReturnType))
        {
            return await next(invocationContext);
        }

        var connectionId = invocationContext.Context.ConnectionId;
        if (connectionId is null)
        {
            return await next(invocationContext);
        }

        var activeStreams = _activeStreams.AddOrUpdate(
            connectionId,
            1,
            (_, current) => current + 1);

        if (activeStreams > _maxConcurrentStreams)
        {
            Decrement(connectionId);
            throw new HubException($"The connection is limited to {_maxConcurrentStreams} concurrent streaming invocations.");
        }

>         try

Reviewer: I had to look at this a few times, but I think the limit would never be honored.

The count is decremented the moment the stream starts delivering, not when the stream actually finishes.
Every stream increments the counter, then decrements it almost instantly as soon as the stream object is returned. So by the time a second stream request comes in, the first one has already decremented. The counter is back to zero even though the first stream is still actively running.

So it seems for example, we could start 100 concurrent streams even though there is a limit of 1, because the counter never stays elevated long enough to block as intended, so no actual concurrency control.

@BrennanConroy, what would be ther correct pattern to suggest and document here?  Also, should this be using `InvokeStreamAsync` as opposed to `InvokeMethodAsync`?

Reviewer: @BrennanConroy, could you weigh in on the correct pattern here for above?  It is not clear to me what it should be.

Reviewer 2: My initial thought would have been code in the Hub itself to handle this: (pseudo code) e.g.
```
class MyHub : Hub
{
    public IAsyncEnumerable<string> Stream()
    {
        return WithLimit(GetStream());
    }

    private async IAsyncEnumerable<string> WithLimit(IAsyncEnumerable<string> originalStream)
    {
        // initial limit check to throw before wrapping if at limit
        if (...) throw;

        IncrementStreamCount();

        try
        {
            await foreach (var item in originalStream)
            {
                yield return item;
            }
        }
        finally
        {
            DecrementStreamCount();
        }
    }
}
```

The hub filter is an interesting idea, unfortunately you would need to use some reflection to achieve the same behavior:

<details>
<summary>Reflection IHubFilter Pseudo Code</summary>

```
public sealed class StreamConcurrencyFilter : IHubFilter
{
    private readonly ConcurrentDictionary<string, int> _activeStreams = new();
    private readonly int _maxConcurrentStreams;

    public StreamConcurrencyFilter(int maxConcurrentStreams = 1)
    {
        if (maxConcurrentStreams < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(maxConcurrentStreams));
        }

        _maxConcurrentStreams = maxConcurrentStreams;
    }

    public async ValueTask<object?> InvokeMethodAsync(
        HubInvocationContext invocationContext,
        Func<HubInvocationContext, ValueTask<object?>> next)
    {
        var streamItemType = GetStreamItemType(invocationContext.HubMethod.ReturnType);
        if (streamItemType is null)
        {
            return await next(invocationContext);
        }

        var connectionId = invocationContext.Context.ConnectionId;

        // Reserve a slot before invoking the method. If the cap is exceeded,
        // release the slot and fail the invocation so no stream is started.
        if (_activeStreams.AddOrUpdate(connectionId, 1, (_, current) => current + 1)
            > _maxConcurrentStreams)
        {
            Decrement(connectionId);
            throw new HubException(
                $"The connection is limited to {_maxConcurrentStreams} concurrent streaming invocations.");
        }

        object? result;
        try
        {
            result = await next(invocationContext);
        }
        catch
        {
            Decrement(connectionId);
            throw;
        }

        if (result is null)
        {
            Decrement(connectionId);
            return null;
        }

        // Wrap the result so the slot is released only after the client finishes
        // consuming the stream (or the connection is aborted).
        return WrapMethod
            .MakeGenericMethod(streamItemType)
            .Invoke(this, new[] { result, connectionId });
    }

    private static readonly MethodInfo WrapMethod = typeof(StreamConcurrencyFilter)
        .GetMethod(nameof(Wrap), BindingFlags.Instance | BindingFlags.NonPublic)!;

    private object Wrap<T>(object result, string connectionId)
    {
        return result switch
        {
            IAsyncEnumerable<T> asyncEnumerable => TrackAsyncEnumerable(asyncEnumerable, connectionId),
            ChannelReader<T> channelReader => TrackChannelReader(channelReader, connectionId),
            _ => ReleaseAndReturn(result, connectionId),
        };
    }

    private async IAsyncEnumerable<T> TrackAsyncEnumerable<T>(
        IAsyncEnumerable<T> source,
        string connectionId,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        try
        {
            await foreach (var item in source.WithCancellation(cancellationToken))
            {
                yield return item;
            }
        }
        finally
        {
            Decrement(connectionId);
        }
    }

    private ChannelReader<T> TrackChannelReader<T>(ChannelReader<T> source, string connectionId)
    {
        var channel = Channel.CreateUnbounded<T>(new UnboundedChannelOptions
        {
            SingleReader = true,
            SingleWriter = true,
        });

        _ = Task.Run(async () =>
        {
            try
            {
                await foreach (var item in source.ReadAllAsync())
                {
                    await channel.Writer.WriteAsync(item);
                }

                channel.Writer.TryComplete();
            }
            catch (Exception ex)
            {
                channel.Writer.TryComplete(ex);
            }
            finally
            {
                Decrement(connectionId);
            }
        });

        return channel.Reader;
    }

    private object ReleaseAndReturn(object result, string connectionId)
    {
        Decrement(connectionId);
        return result;
    }

    private static Type? GetStreamItemType(Type returnType)
    {
        if (returnType is null)
        {
            return null;
        }

        if (returnType.IsGenericType)
        {
            var def = returnType.GetGenericTypeDefinition();
            if (def == typeof(IAsyncEnumerable<>) || def == typeof(ChannelReader<>))
            {
                return returnType.GetGenericArguments()[0];
            }

            if (def == typeof(Task<>) || def == typeof(ValueTask<>))
            {
                return GetStreamItemType(returnType.GetGenericArguments()[0]);
            }
        }

        return null;
    }

    private void Decrement(string connectionId)
    {
        while (_activeStreams.TryGetValue(connectionId, out var current))
        {
            if (current <= 1)
            {
                if (_activeStreams.TryRemove(new KeyValuePair<string, int>(connectionId, current)))
                {
                    return;
                }
            }
            else if (_activeStreams.TryUpdate(connectionId, current - 1, current))
            {
                return;
            }
        }
    }
}
```

</details>

        var current = _activeStreams.AddOrUpdate(
            connectionId,
            addValue: 1,
            updateValueFactory: (_, count) => count + 1);

        if (current > MaxConcurrentStreams)
        {
            Decrement(connectionId);
            throw new HubException(
                $"The connection is limited to {MaxConcurrentStreams} concurrent streaming invocations.");
        }

        try
        {
            await foreach (var item in stream.WithCancellation(cancellationToken))
            {
                yield return item;
            }
        }
        finally
        {
            Decrement(connectionId);
        }
    }

    private static void Decrement(string connectionId)
    {
        while (_activeStreams.TryGetValue(connectionId, out var current))
        {
            if (current <= 1)
            {
                if (_activeStreams.TryRemove(new KeyValuePair<string, int>(connectionId, current)))
                {
                    return;
                }
            }
            else if (_activeStreams.TryUpdate(connectionId, current - 1, current))
            {
                return;
            }
        }
    }
}
```

> The key point is that `WithLimit` wraps the original `IAsyncEnumerable<T>` and holds the counter elevated for the *full lifetime of the stream*, not just until the first item is yielded.

Reviewer 2: We should also mention that `ChannelReader<T>` could have a similar helper method applied to it (and should use the same dictionary).

To enforce a per-connection limit on streaming invocations, wrap the stream inside the hub method itself using a private helper that increments a counter before yielding items and decrements it in a `finally` block:

```csharp
using System.Collections.Concurrent;
using System.Runtime.CompilerServices;

public class StreamingHub : Hub
{
    // Track the number of active streams per connection.
    private static readonly ConcurrentDictionary<string, int> _activeStreams = new();
    private const int MaxConcurrentStreams = 2;

    public IAsyncEnumerable<int> Counter(
        int count,
        int delay,
        CancellationToken cancellationToken)
    {
        return WithLimit(GetCounter(count, delay, cancellationToken));
    }

    private async IAsyncEnumerable<int> GetCounter(
        int count,
        int delay,
        [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        for (var i = 0; i < count; i++)
        {
            cancellationToken.ThrowIfCancellationRequested();
            yield return i;
            await Task.Delay(delay, cancellationToken);
        }
    }

>     private async IAsyncEnumerable<int> WithLimit(

Reviewer 2: Should this use `IAsyncEnumerable<T>`?

Author: This came from the suggestion by @wadepickett. Whats your thought on this @BrennanConroy. Do we make it generic of type T?

Reviewer: @BrennanConroy,  I think this is the very last item/question to resolve.

Reviewer 2: I think we should make it generic. Also pass in the ConnectionId as an argument to make it more obvious how someone could extract this method into a service or utility class outside of the Hub.

        IAsyncEnumerable<int> stream,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var connectionId = Context.ConnectionId;

        var current = _activeStreams.AddOrUpdate(
            connectionId,
            addValue: 1,
            updateValueFactory: (_, count) => count + 1);

        if (current > MaxConcurrentStreams)
        {
            _activeStreams.AddOrUpdate(
                connectionId,
                addValue: 0,
                updateValueFactory: (_, count) => Math.Max(0, count - 1));

            throw new HubException(
                $"The connection is limited to {MaxConcurrentStreams} concurrent streaming invocations.");
        }

        try
        {
            await foreach (var item in stream.WithCancellation(cancellationToken))
            {
                yield return item;
            }
        }
        finally
        {
            _activeStreams.AddOrUpdate(
                connectionId,
                addValue: 0,

>                 updateValueFactory: (_, count) => Math.Max(0, count - 1));

Reviewer 2: I think this also needs to remove itself from the dictionary to avoid a memory leak
```
_activeStreams.TryRemove(new (connectionId, 0));
```

Above as well in the error path.

Author: I have fixed this in the code snippet. Made the decrement logic a private function that handles decrement/removal from dictionary depending on how many connections are left.

        }
    }
}
```

The key point is that `WithLimit` wraps the original `IAsyncEnumerable<T>` and holds the counter elevated for the *full lifetime of the stream*, not just until the first item is yielded.
The `finally` block runs only when the client finishes consuming the stream, cancels it, or the connection drops.

> [!NOTE]
> The `_activeStreams` dictionary is `static` so it is shared across all hub instances for

> a given connection. If you prefer DI-managed state, register a singleton service that owns

Reviewer 2: ```suggestion
> The `_activeStreams` dictionary is `static` so it is shared across all hub instances. If you prefer DI-managed state, register a singleton service that owns
```

Author: Updated now

### Keyed services support in dependency injection

> The keyed services mechanism allows you to register and retrieve dependency injection services by using keys. A service is associated with a key by calling the  xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddKeyedSingleton%2A method to register it. As an alternative, you can call the `AddKeyedScoped` or `AddKeyedTransient` method.

Reviewer: ```suggestion
The keyed services mechanism allows you to register and retrieve dependency injection services by using keys. A service is associated with a key by calling the  <xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddKeyedSingleton%2A> method to register it. As an alternative, you can call the `AddKeyedScoped` or `AddKeyedTransient` method.
```
Replaced missing angle brackets.

> You access a registered service by specifying the key with the [FromKeyedServices] attribute. The following code shows how to use keyed services:

Reviewer: ```suggestion
You access a registered service by specifying the key with the [[FromKeyedServices](xref:Microsoft.Extensions.DependencyInjection.FromKeyedServicesAttribute)] attribute. The following code shows how to use keyed services:
```
Making this a link.

:::code language="csharp" source="~/../AspNetCore.Docs.Samples/signalr/hubs/KeyedSvsHub/Program.cs" highlight="5-6,34,39":::

## Limit per-connection streaming invocations

`<xref:Microsoft.AspNetCore.SignalR.HubOptions.MaximumParallelInvocationsPerClient> controls the number of non-streaming hub method invocations a client can run in parallel before they are queued. It does **not** apply to streaming hub invocations. Streaming invocations are intentionally excluded because they are expected to be long-running and concurrent, so a client can start any number of concurrent streams regardless of that setting.

To enforce a per-connection limit on streaming invocations, wrap the stream inside the hub method itself using a private helper that increments a counter before yielding items and decrements it in a `finally` block:

```csharp
using System.Collections.Concurrent;
using System.Runtime.CompilerServices;

public class StreamingHub : Hub
{
    // Track the number of active streams per connection.

>     private static readonly ConcurrentDictionary _activeStreams = new();

Reviewer: ```suggestion
    private static readonly ConcurrentDictionary<string, int> _activeStreams = new();
```

I think everything in angle brackets may have been stripped accidently during editing?  The generic type parameters on IAsyncEnumerable, ConcurrentDictionary, etc. got stripped out. Compare your code block to the [official sample](https://github.com/dotnet/aspnetcore/blob/7f0d5e98459ff1e14bf1859e4409abe7ccd9d5a6/src/SignalR/samples/SignalRSamples/Hubs/Streaming.cs#L13). For example IAsyncEnumerable needs to be IAsyncEnumerable<int>, and ConcurrentDictionary needs to be ConcurrentDictionary<string, int>. Without these the sample won't compile.

    private const int MaxConcurrentStreams = 2;

>     public IAsyncEnumerable Counter(

Reviewer: ```suggestion
    public IAsyncEnumerable<int> Counter(
```
Bracket items seem to have been removed.

        int count,
        int delay,
        CancellationToken cancellationToken)
    {
        return WithLimit(GetCounter(count, delay, cancellationToken));
    }

>     private async IAsyncEnumerable GetCounter(

Reviewer: ```suggestion
    private async IAsyncEnumerable<int> GetCounter(
```
Put missing generic type params back in.

        int count,
        int delay,
        [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        for (var i = 0; i < count; i++)
        {
            cancellationToken.ThrowIfCancellationRequested();
            yield return i;
            await Task.Delay(delay, cancellationToken);
        }
    }

>     private async IAsyncEnumerable WithLimit(

Reviewer: ```suggestion
    private async IAsyncEnumerable<int> WithLimit(
```
Putting the missing generic type param back in.

>         IAsyncEnumerable stream,

Reviewer: ```suggestion
        IAsyncEnumerable<int> stream,
```
I put the missing generic type params back in.

                updateValueFactory: (_, count) => Math.Max(0, count - 1));

            throw new HubException(
                $"The connection is limited to {MaxConcurrentStreams} concurrent streaming invocations.");
        }

        try
        {
            await foreach (var item in stream.WithCancellation(cancellationToken))
            {
                yield return item;
            }
        }
        finally
        {
            _activeStreams.AddOrUpdate(
                connectionId,
                addValue: 0,
                updateValueFactory: (_, count) => Math.Max(0, count - 1));
        }
    }
}
```

The key point is that `WithLimit` wraps the original `IAsyncEnumerable<T>` and holds the counter elevated for the *full lifetime of the stream*, not just until the first item is yielded.
The `finally` block runs only when the client finishes consuming the stream, cancels it, or the connection drops.

> [!NOTE]
> The `_activeStreams` dictionary is `static` so it is shared across all hub instances for
a given connection. If you prefer DI-managed state, register a singleton service that owns
> the dictionary and inject it into the hub constructor.

> ### Keyed services support in Dependency Injection

Reviewer: This "Keyed services support.." section now appears twice.  This one from line 352 - 358 needs to be removed so the duplicate no longer exists, and to get rid of an extra line space.
